### PR TITLE
Enable multiple spectra to share the same image without data duplication

### LINKED ISSUE #740: [FEAT] Mosviz: allow mosviz to load a *single* image which is used for all of the rows instead of separate image for each row
This is a companion/follow-on to #739 - the idea here would be to have a mode where instead of each spectrum getting its own cutout image, a *single* image could be used for all the rows.  Combined with #739, this could be basically transparent, where instead of swapping in a different image when you switch to a different object, it just moves to a different spot in the same image.

The interface would just be
```
mosviz.load_images(a_single_nddata_object_instead_of_a_list_of_them)
```
the one catch is that unlike the current approach, in this case it's critical that the loaded image have a wcs, since we can't rely on the assumption that it's a pixel-perfect cutout - the wcs is necessary to match the spectrum's spatial part to the image.  

This might be less of a change that it first appears to the current idiom, in that we could basically just have the above be syntactic sugar for 
```
mosviz.load_images([image]*number_of_rows)
```
and then as long as mosviz does not *re-load* an image it's already loaded once, it would all work well.  Then the user can even use, say, two different images with 5 rows, as long as they can manually keep track of which image goes with which row.

(Caveat: I suppose we *could* have some optional columns in the table that give the center of the image in *pixel* coordinates and the corresponding image... but that requires more thought than what I propose above, so probably is best done as a follow-on later).

### PR DESCRIPTION
**Pull request by rosteen**
### Description
Enables users to input a single image filename string (and optionally a single associated label) to the Mosviz `load_data()` method that will be associated with all spectra in a list of input spectra. Also sets a couple currently-unused attributes (e.g. `_shared_image` in the table viewer) that will be useful later for addressing #739. Can be tested with the updated MosvizExample notebook, which now passes in the single cutout image instead of duplicating it into a list in the notebook before calling `load_data`.

Closes #740 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:263]:
```suggestion
        Parse and load a 2D image. ``CCDData`` objects require a unit be defined
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:269]: That is a dangerous assumption, but perhaps out of scope.

**Review by pllim — COMMENTED** [review]
So, what I am supposed to be looking at? I run the example notebook. The value of `images` is `.../mosviz_nirspec_data_0.3/level3/mosviz_cutouts/227.fits`. `mosviz.load_data(spectra_1d, spectra_2d, images=images, images_label="Cutout 227")` did not error out and gave me the `BUNIT` warning, but I don't see any image being displayed. Did I do something wrong?

Screenshot 2021-08-04 162342

**Review by pllim — APPROVED** [review]
Doh, nevermind... Turns out I have to manually select "Cutout 227" for it to display. Works for me now.

**Comment by pllim** [comment]
Is it possible to add tests? 😉

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (rosteen) [reply jdaviz/configs/mosviz/plugins/parsers.py:269]: I agree 100%, but yes it's out of scope.

**Comment by rosteen** [comment]
I'll add a test or two today, I usually drag my feet on that and do it after I make the PR. Bad habit.

**Comment by pllim** [comment]
I do wonder though... when there is only one image, can the image viewer part just display it instead me having to go manually check a box? 💭

**Comment by rosteen** [comment]
Well, I think the intent is eventually be able to do something like have one image for the first 6 rows, a second image for the next 6 rows, etc. Not currently possible, but I don't want to preclude that possibility.

File: jdaviz/configs/mosviz/tests/test_helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_helper.py:213]: Simpler like this?
```suggestion
    spectra1d = [spectrum1d] * 3
    spectra2d = [spectrum2d] * 3
```

File: jdaviz/configs/mosviz/tests/test_helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_helper.py:220]: Do we also need to check the length of `qtable`?

File: jdaviz/configs/mosviz/tests/test_helper.py
Comment (rosteen) [reply jdaviz/configs/mosviz/tests/test_helper.py:213]: Yes, but I was copying from an existing test. Maybe I'll go ahead and change all of them...

File: jdaviz/configs/mosviz/tests/test_helper.py
Comment (rosteen) [reply jdaviz/configs/mosviz/tests/test_helper.py:220]: Wouldn't hurt, but really maybe I should add a separate test of `to_table`. Apparently there's only one for `to_csv` (probably my fault).

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:255]: Maybe elaborate on _how_ this non-zero number will affect the display.

**Comment by rosteen** [comment]
Fun fact, if you want `pytest` to run your test, you have to name it `test_`...

I think this is good to go now, added a little more clarity to the docstring you mentioned @pllim .

**Comment by pllim** [comment]
Re: fun fact -- Not necessarily just the prefix, but it does have to be discoverable: https://docs.pytest.org/en/6.2.x/goodpractices.html 😸

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:254]: As an outsider, it is still unclear to me. What is the difference if I pass in `share_image=1` and `share_image=2`?

Comment (rosteen) [reply jdaviz/configs/mosviz/plugins/parsers.py:254]: As a user, you don't pass in that argument at all. The user calls one of the `load_` methods, which in turn call the parsers on the backend. This is an argument that gets populated automatically based on the input to the `load_data` method, so the docstring is really developer-facing IMO.

**Comment by pllim** [comment]
There appears to be a couple of untested logic blocks... Is it possible to test those too?

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (rosteen) [reply jdaviz/configs/mosviz/plugins/parsers.py:254]: Still, I guess you're a developer, so it should be useful to you...

**Comment by rosteen** [comment]
Yes, but doing so has raised an issue (predating this PR) that I need to fix now. See how dangerous testing is!

**Comment by rosteen** [comment]
I created #763 as a followup to generally improve the Mosviz test coverage not specific to this functionality. I think that this can be merged at this point.

**Comment by rosteen** [comment]
@eteq I'll give you the opportunity to give a final thumbs up on this before I merge, since it addresses your feature request.

**Comment by rosteen** [comment]
Merging, we have two approvals and the window for objections has passed (according to me, anyway).

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (eteq) [reply jdaviz/configs/mosviz/plugins/parsers.py:254]: FWIW, I also was confused what the meaning of "not 0" is from the docstring.

**Comment by eteq** [comment]
I give a post-facto :+1:. I left one comment above that I think there's an unclear docstring, but that would be a trivial follow-on PR whenever you have a chance @rosteen .

File: jdaviz/configs/mosviz/helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/helper.py:200]: Belated comment: `share_image` here is not documented.

Comment (pllim) [reply jdaviz/configs/mosviz/helper.py:200]: See #784

**Pull request merged**
